src/pixelator/analysis/normalization/__init__.py

```diff
@@ -0,0 +1,68 @@
+"""Functions for the colocalization analysis in pixelator.
+
+Copyright © 2024 Pixelgen Technologies AB.
+"""
+
+from typing import List, Union
+
+import numpy as np
+import pandas as pd
+from sklearn.decomposition import PCA
+from sklearn.mixture import GaussianMixture
+
+
+def _limma(pheno, exprs, rcond=1e-8):
+    """Linear regression to remove confounding factors from expression data."""
+    design_matrix = np.column_stack((np.ones((len(pheno), 1)), pheno))
+    coefficients, res, rank, s = np.linalg.lstsq(design_matrix, exprs, rcond=rcond)
+    beta = coefficients[1:]  # remove intercept term
+    return exprs - design_matrix[:, 1:].dot(beta)
+
+
+def _get_baseline_expression(dataframe: pd.DataFrame, axis=0):
+    """Fit a double gaussian distribution to the expression data and return mean of the first gaussian as baseline."""
+    baseline = pd.Series(index=dataframe.index if axis == 0 else dataframe.columns)
+    scores = pd.Series(index=dataframe.index if axis == 0 else dataframe.columns)
+    gmm = GaussianMixture(n_components=2, max_iter=1000, random_state=0)
+    if axis == 0:
+        for i in dataframe.index:
+            marker_data = dataframe.loc[i, :].to_frame()
+            gmm = gmm.fit(marker_data)
+            baseline[i] = np.min(gmm.means_)
+            scores[i] = np.abs(gmm.means_[1] - gmm.means_[0]) / np.sum(gmm.covariances_)
+    elif axis == 1:
+        for i in dataframe.columns:
+            marker_data = dataframe.loc[:, i].to_frame()
+            gmm = gmm.fit(marker_data)
+            baseline[i] = np.min(gmm.means_)
+            scores[i] = np.abs(gmm.means_[1] - gmm.means_[0]) / np.sum(gmm.covariances_)
+    return baseline, scores
+
+
+def dsb_normalize(
+    raw_expression: pd.DataFrame, isotype_controls: Union[List, None] = None
+):
+    """empty-droplet-free method as implemented in the dsb package.
```

Comment:
I think we should add a reference to the original package here.

Reply:
Mulè, M.P., Martins, A.J. & Tsang, J.S. Normalizing and denoising protein expression data from droplet-based single cell profiling. Nat Commun 13, 2099 (2022).
https://doi.org/10.1038/s41467-022-29356-8